protocol.md

```diff
@@ -870,6 +902,12 @@ export interface TextDocumentClientCapabilities {
 ```typescript
 interface ClientCapabilities {
 	/**
+	 * The client supports receiving the result solely through $/partialResult notifications for requests from the client to the server.
+	 * If true, the server should send all data through $/partialResult and send null as the result in the final response.
```

Comment:
> All methods should behave consistently and either all use $/partialResult or none. This makes the receiver code less complicated.

I don't think the "this makes the receiver code less complicated" part is true. Clients will need to support servers that don't support streaming capabilities, so they will need to write this code anyway.

> What if a method doesn't want to use Partial Result? 

I agree that it does seem wasteful that "simple" methods have to send two responses (one partial result and then a final result). This would be an argument for either (1) per method streaming capability or (2) adding streaming versions of each relevant method to the spec.

> I worry about jsonrpc2 implementations making it difficult to construct a consistent order of when it receives a notification and a response, since we are now dictating an ordering requirement on responses/notifications that never existed before.

The concurrency issue does seem like a potential problem if the server's json rpc implementation doesn't guarantee ordering (e.g. server code sends partial followed by final, but due to goroutine scheduling the final is sent before the partial).

It seems wasteful to send all the data again in the final response, but that is the simplest solution.

***

At this point I am actually on the fence about whether the server should be required to send the data in the final response or required not to send it. What I can say is that we shouldn't leave the decision up to the server implementation; we should think about the tradeoffs now and the spec should require one behavior or the other.

Here is an alternative proposal to think about

**$/streamRequest**
JSON RPC notification sent from client to server to initiate a streaming request.

Parameters:
- ID for this request. Note that this is an application level param, not an ID assigned by JSON RPC.
- Name of another LSP method
- Parameters for that LSP method

**$/partialResult**
JSON RPC notification sent from server to client to send results back for a $/streamRequest.

Parameters:
- ID from the associated $/streamRequest
- JSON patch
- a boolean to indicate if this is the final notification and the streaming request is finished

Thoughts?